src/java.base/share/classes/sun/security/provider/SecureRandom.java

```diff
@@ -154,15 +162,21 @@ private void init(byte[] seed) {
      * @param seed the seed.
      */
     @Override
-    public synchronized void engineSetSeed(byte[] seed) {
-        if (state != null) {
-            digest.update(state);
-            for (int i = 0; i < state.length; i++) {
-                state[i] = 0;
+    public void engineSetSeed(byte[] seed) {
+        objLock.lock();
+        try {
+            if (state != null) {
+                digest.update(state);
+                for (int i = 0; i < state.length; i++) {
+                    state[i] = 0;
+                }
             }
+            state = digest.digest(seed);
+            remCount = 0;
+            clearStateOnCheckpoint = false;
```

Comment:
I'm not clear on why we need this boolean.  Why does the code say not to clear this Random on a checkpoint if someone has called `engineSetSeed`?

Reply:
By default, SHA1PRNG implementation depends on the initial seed and is not reseeded internally. The random sequence is a function of initial seed. So, an application could expect a particular sequence on depends on the seed provided. This is a reason we can not reseed secure random if the application provides its own seed. In case of application does not provide seed, the seed is auto-assigned by implementation. It means the application is not depends on particular random sequence. It means we can re-seed such secure randoms during checkpoint/restore.
Secure Randoms, seeded by the application should be handled by this application layer because of only application knows How to handle it.
So, we'll re-seed secure randoms only in case of we sure it does not change application logic

Reply:
That makes a lot of sense - thanks for clarifying!